File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.js
function solution(participant, completion) {
    let rst = '';
    // js에서 객체 생성에는 hashmap 보다 성능이 좋은 방식을 활용함
    const hash = {};
    // 1. init hash
    for(let i=0; i<participant.length;i++){
        const currentKey = participant[i];
        if(hash[currentKey]){
            hash[currentKey]++;
        } else {
            hash[currentKey] = 1;
        }
    }
    
    // 2. loop 순회, 카운트 빼기
    for(let i=0; i<completion.length; i++){
        const currentKey = completion[i];
        if(hash[currentKey] === 0) {
            rst = currentKey;
            break;
        } else {
            hash[currentKey]--;
        }
    }
    // 3. 완주 못한 사람 확인
    for(let key in hash){
        if(hash[key] > 0){
            rst = key;
            break;
        }
    }
    
    return rst;
}

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.py
# participant <= 100,000
def solution(participant, completion):
    ans = ''
    # hashmap
    d = {}
    for p in participant:
        if p in d.keys():
            d[p] +=1
        else:
            d[p] = 1
    
    # check hashmap
    # O(N)
    for p in completion:
        d[p] -= 1
    
    for k,v in d.items():
        if v >=1:
            ans = k
            break
    
    return ans